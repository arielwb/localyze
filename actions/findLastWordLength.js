export const findLastWordLength = (string) => {
  const getLastWord = /(\w+)\s*$/
  const lastWord = getLastWord.exec(string)[1]

  return lastWord.length
}
