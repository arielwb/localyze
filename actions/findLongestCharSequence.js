export const findLongestCharSequence = (word) => {
  let longestCharSequence = ''
  let currentCharSequence = ''
  const characters = [...word]

  characters.forEach(char => {
    if (!longestCharSequence) {
      longestCharSequence = char
      currentCharSequence = char

      return
    }

    if (currentCharSequence[0] !== char) {
      if (currentCharSequence.length > longestCharSequence.length) {
        longestCharSequence = currentCharSequence
      }

      currentCharSequence = char
    } else {
      currentCharSequence += char
    }
  })

  return longestCharSequence
}
