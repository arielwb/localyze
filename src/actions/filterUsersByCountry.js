export const sampleUsers = [
  { name: 'Maddy', country: 'Brazil' },
  { name: 'Nate', country: 'Netherlands' },
  { name: 'Cassie', country: 'Germany' },
  { name: 'Fezco', country: 'Norway' },
  { name: 'Rue', country: 'France' },
  { name: 'Jules', country: 'France' },
  { name: 'Elliot', country: 'Poland' },
  { name: 'Lexi', country: 'Poland' }
]

export const filterUsersByCountry = (setState, users, countryFilter) => {
  if (!countryFilter) {
    setState(users)
  } else {
    const filteredCountries = users.filter((user) => {
      const country = user.country.toLowerCase()
      const filter = countryFilter.toLowerCase()

      return country.includes(filter)
    })

    setState(filteredCountries)
  }
}
