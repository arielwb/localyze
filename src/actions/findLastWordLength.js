export const findLastWordLength = (string) => {
  if (string.length < 1 || string.length > 104) {
    return null
  }

  const getLastWord = /(\w+)\s*$/
  const lastWord = getLastWord.exec(string)[1]

  return lastWord.length
}
