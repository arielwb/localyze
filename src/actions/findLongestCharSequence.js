export const findLongestCharSequence = (word) => {
  let longestCharSequence = ''
  let currentCharSequence = ''
  const characters = [...word]

  characters.forEach(char => {
    if (!longestCharSequence) {
      longestCharSequence = char
      currentCharSequence = char

      return
    }

    currentCharSequence = currentCharSequence[0] !== char
      ? char
      : currentCharSequence + char

    if (currentCharSequence.length > longestCharSequence.length) {
      longestCharSequence = currentCharSequence
    }
  })

  return longestCharSequence
}
