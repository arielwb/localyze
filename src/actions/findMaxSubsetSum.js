const sum = (numbers) => numbers.reduce((currentSum, value) => currentSum + value, 0)

const findMaxNumberIndex = (numbers) => {
  let maxNumber = numbers[0]
  let maxNumberIndex = 0

  numbers.forEach((num, index) => {
    if (num > maxNumber) {
      maxNumber = num
      maxNumberIndex = index
    }
  })

  return maxNumberIndex
}

export const findMaxSubsetSum = (numbers) => {
  if (!numbers.length) return []

  const negativeNumbersIndex = []
  let prevIndex
  let prevSum = 0
  let startIndex = 0
  let endIndex = numbers.length - 1

  numbers.forEach((num, index) => num < 0 && negativeNumbersIndex.push(index))

  if (negativeNumbersIndex.length === numbers.length) {
    const maxNegativeNumberIndex = findMaxNumberIndex(numbers)

    return [maxNegativeNumberIndex, maxNegativeNumberIndex]
  }

  negativeNumbersIndex.forEach((negativeNumIndex) => {
    const range = numbers.slice(prevIndex, negativeNumIndex)
    const rangeSum = sum(range)

    if (prevIndex) {
      const positiveRangeSum = range.length === 1 ? rangeSum : rangeSum - numbers[prevIndex]
      const totalSum = prevSum + rangeSum

      if (positiveRangeSum > totalSum) {
        startIndex = prevIndex + 1
        prevSum = positiveRangeSum
        endIndex = negativeNumIndex - 1
      } else if (totalSum > prevSum) {
        prevSum = totalSum
        endIndex = negativeNumIndex - 1
      }
    } else {
      prevSum = rangeSum
      endIndex = negativeNumIndex - 1
    }

    prevIndex = negativeNumIndex
  })

  const lastNegativeIndex = negativeNumbersIndex.pop()

  if (lastNegativeIndex !== numbers.length - 1) {
    const lastRange = numbers.slice(lastNegativeIndex)
    const lastRangeSum = sum(lastRange)
    const lastPositiveRangeSum = lastRangeSum - numbers[lastNegativeIndex]

    if (lastPositiveRangeSum > prevSum) {
      startIndex = lastNegativeIndex + 1
      endIndex = numbers.length - 1
    } else if (prevSum + lastRangeSum > prevSum) {
      endIndex = numbers.length - 1
    }
  }

  return [startIndex, endIndex]
}
