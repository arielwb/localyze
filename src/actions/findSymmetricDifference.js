const getUniqueItems = (reference, set) => reference.filter(number => !set.has(number))

export const findSymmetricDifference = (sample1, sample2) => {
  if (sample1.length > 100 || sample2.length > 100) {
    return null
  }

  const set1 = new Set(sample1)
  const set2 = new Set(sample2)

  return [...getUniqueItems(sample1, set2), ...getUniqueItems(sample2, set1)]
}
