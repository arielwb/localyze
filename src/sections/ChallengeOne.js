import React, { useState } from 'react'
import { Header, Input, OutputBox } from '../components'
import { findLongestCharSequence } from '../actions/findLongestCharSequence'

export const ChallengeOne = () => {
  const [result, setResult] = useState()
  const [userInput, setUserInput] = useState('')

  const onSubmit = (event) => {
    const longestCharSequence = findLongestCharSequence(userInput)
    setResult(longestCharSequence)
  }

  return (
    <>
      <Header id={1}>
        Given a word, write an algorithm that returns the sequence of a specific
        character that appears the most in sequence.
        If there are many sequences like that, return the first longest sequence.
      </Header>
      <Input
        label='Ex: Pressuuuuuure'
        name='challenge1'
        type='text'
        value={userInput}
        onSubmit={onSubmit}
        onChange={(event) => setUserInput(event.target.value)}
      />
      <OutputBox>{result}</OutputBox>
    </>
  )
}
