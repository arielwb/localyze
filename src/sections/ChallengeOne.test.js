import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react'
import { ChallengeOne } from './ChallengeOne'
import * as actions from '../actions/findLongestCharSequence'

describe('ChallengeOne', () => {
  it('should match the snapshot', () => {
    const tree = renderer
      .create(<ChallengeOne />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call findLongestCharSequence function on submit', () => {
    const findLongestCharSequenceSpy = jest.spyOn(actions, 'findLongestCharSequence')
    render(<ChallengeOne />)
    const button = screen.getByTestId('submit-button')
    const input = screen.getByTestId('input')

    fireEvent.change(input, { target: { value: 'new value' } })
    fireEvent.submit(button)

    expect(findLongestCharSequenceSpy).toHaveBeenCalled()
  })
})
