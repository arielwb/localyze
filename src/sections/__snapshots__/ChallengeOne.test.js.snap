// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengeOne should match the snapshot 1`] = `
Array [
  <h1>
    Challenge 
    1
  </h1>,
  <p>
    Given a word, write an algorithm that returns the sequence of a specific character that appears the most in sequence. If there are many sequences like that, return the first longest sequence.
  </p>,
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="challenge1"
    >
      Ex: Pressuuuuuure
    </label>
    <input
      data-testid="input"
      name="challenge1"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      data-testid="submit-button"
      type="submit"
    >
      confirm
    </button>
  </form>,
  <pre />,
]
`;
